src/Schema/Directives/Fields/CanDirective.php

```diff
@@ -36,25 +37,64 @@ public function handleField(FieldValue $value, \Closure $next)
         return $next(
             $value->setResolver(
                 function () use ($resolver) {
-                    /** @var Authorizable $user */
                     $user = auth()->user();
+                    $gate = resolve(Gate::class);
+                    $args = $this->getArguments();
 
-                    if (!$user) {
-                        throw new AuthenticationException('Not authenticated to access this field.');
-                    }
-
-                    $model = $this->getModelClass();
-                    $policies = $this->directiveArgValue('if');
-
-                    collect($policies)->each(function (string $policy) use ($user, $model) {
-                        if (!$user->can($policy, $model)) {
-                            throw new AuthorizationException('Not authorized to access this field.');
-                        }
+                    $this->getAbilities()->each(function (string $ability) use ($args, $gate, $user) {
+                        $this->validate($user, $gate, $ability, $args);
                     });
 
-                    return call_user_func_array($resolver, func_get_args());
+                    return \call_user_func_array($resolver, \func_get_args());
                 }
             )
         );
     }
+
+    /**
+     * Get the ability argument.
+     *
+     * @return Collection
+     */
+    protected function getAbilities(): Collection
+    {
+        return collect(
+            $this->directiveArgValue('ability') ??
+            $this->directiveArgValue('if')
+        );
+    }
+
+    /**
+     * Get the arguments passing to `Gate::check`.
+     *
+     * @throws \Exception
+     *
+     * @return array
+     */
+    protected function getArguments(): array
+    {
+        $modelClass = $this->getModelClass();
+        $args = (array) $this->directiveArgValue('args');
+
+        array_unshift($args, $modelClass);
```

Comment:
What is going on here? Please elaborate with a comment

Reply:
(added a comment)